src/resolution/resolver.ts

```diff
@@ -79,7 +81,7 @@ const _resolveRequest = (requestScope: interfaces.RequestScope) =>
             result = binding.cache;
             binding.activated = true;
         } else if (binding.type === BindingTypeEnum.Constructor) {
-            result = binding.implementationType;
+            result = binding.implementationType as unknown as T;
```

Comment:
I think we should add breaking changes to solve it. Do you have anything in mind?

We have plans to change typings, but I think we should accept they are not perfect :(

Reply:
> Do you have anything in mind?

It will need some thought for sure !

> I think we should add breaking changes to solve it

That will probably be necessary.

It's only a problem as resolveRequest is typed to return `undefined | T | Promise<T> | (T | Promise<T>`
All other returns are typed as any !

Perhaps better to untype the return for now.

Problems still remain but they were there before this feature.

I will happily look at the typings at a later date.

What I would like to do is to remove all of this code !

```
if (binding.type === BindingTypeEnum.ConstantValue) {
            result = binding.cache;
            binding.activated = true;
        } else if (binding.type === BindingTypeEnum.Function) {
            result = binding.cache;
            binding.activated = true;
        } else if (binding.type === BindingTypeEnum.Constructor) {
            result = binding.implementationType as unknown as T;
        } else if (binding.type === BindingTypeEnum.DynamicValue && binding.dynamicValue !== null) {
            result = invokeFactory(
                "toDynamicValue",
                binding.serviceIdentifier,
                () => (binding.dynamicValue as (context: interfaces.Context) => any)(request.parentContext)
            );
        } else if (binding.type === BindingTypeEnum.Factory && binding.factory !== null) {
            result = invokeFactory(
                "toFactory",
                binding.serviceIdentifier,
                () => (binding.factory as interfaces.FactoryCreator<any>)(request.parentContext)
            );
        } else if (binding.type === BindingTypeEnum.Provider && binding.provider !== null) {
            result = invokeFactory(
                "toProvider",
                binding.serviceIdentifier,
                () => (binding.provider as interfaces.Provider<any>)(request.parentContext)
            );
        } else if (binding.type === BindingTypeEnum.Instance && binding.implementationType !== null) {
            result = resolveInstance(
                binding,
                binding.implementationType,
                childRequests,
                _resolveRequest(requestScope)
            );
        } else {
            // The user probably created a binding but didn't finish it
            // e.g. container.bind<T>("Something"); missing BindingToSyntax
            const serviceIdentifier = getServiceIdentifierAsString(request.serviceIdentifier);
            throw new Error(`${ERROR_MSGS.INVALID_BINDING_TYPE} ${serviceIdentifier}`);
        }
```
and instead `binding.provideValue(context)` and have binding as abstract  with a derivation for each.  Perhaps this would help.

Reply:
Well T is supposed to be the result's type, I understand your feeling, I really want to improve that code's block, but (T | Promise<T> | (T|Promise<T>)[]) is a valid returning type. A constructor binding is suposed to be called with a `T extends Newable<unknown>`, considering that  (T | Promise<T> | (T|Promise<T>)[]) is a valid return type. 

I'm looking forward to improve that code, but I believe achieving changes through small PR is the best way to go. It's easier to revier a small PR, it's easier to solve conflicts so it helps this project to scale. So I prefer to solve this in another PR

Reply:
We can discuss the return type at a later date.  I agree another PR for type changes.